### PR TITLE
Use more strict comparisons if possible in Shopware Components

### PR DESCRIPTION
**Pull request by pascalheidmann**
### 1. Why is this change necessary?
Code quality in `/engine/Shopware/Components`

### 2. What does this change do, exactly?
Use strict comparisons if possible, and replace magic numbers with more vocal consts.

### 4. Please link to the relevant issues (if any).
https://github.com/shopware/shopware/pull/2271

### 6. Checklist

- [x] I have written tests and verified that they fail without my change
- [x] I have squashed any insignificant commits
- [x] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfil them.

### CODE REVIEW

File: engine/Shopware/Components/Api/Resource/PropertyGroup.php [hunk 1]
```diff
@@ -214,7 +214,7 @@ private function preparePropertyData(array $params, $propertyGroup = null)
             }
 
             // Sortmode equals the old article_count sorting?
-            if ($params['sortmode'] == 2) {
+            if ($params['sortmode'] === 2) {
```
Comment (shyim) [inline engine/Shopware/Components/Api/Resource/PropertyGroup.php]: cast maybe?

File: engine/Shopware/Components/DependencyInjection/LegacyPhpDumper.php [hunk 2]
```diff
@@ -68,22 +68,39 @@ class LegacyPhpDumper extends Dumper
     const NON_FIRST_CHARS = 'abcdefghijklmnopqrstuvwxyz0123456789_';
 
     private $definitionVariables;
+
```
Comment (shyim) [inline engine/Shopware/Components/DependencyInjection/LegacyPhpDumper.php]: Don't touch this file. We need the ability to diff it with the class from symfony

Comment (pascalheidmann) [reply engine/Shopware/Components/DependencyInjection/LegacyPhpDumper.php]: I guess PHPstan was quite successful in reformating 😅

File: engine/Shopware/Components/Api/Resource/PropertyGroup.php (hunk 1 above)
Comment (pascalheidmann) [reply engine/Shopware/Components/Api/Resource/PropertyGroup.php]: Good point.
But considering it, it looks like code coverage is missing here

Comment (pascalheidmann) [reply engine/Shopware/Components/Api/Resource/PropertyGroup.php]: After looking into it: sortmode had always to be `integer` according to https://developers.shopware.com/developers-guide/rest-api/api-resource-property-group/

**Comment by mitelg** [comment]
are all changes saved by tests? we do not have strict type mode, so this would be valid code:
```
$nullableValue = false;
if ($nullableValue == null) { // <- true!
}
```
we have to ensure, that we don't change the current behaviour or create errors through restricting the types

**Comment by pascalheidmann** [comment]
@mitelg some had crashed initially and had to be casted.
But most are comparisons between strings anyway where something other than a string already won't match.
The few occasions I found (mysql queries are horrible as they are always strings!) I casted for now.

File: engine/Shopware/Components/Plugin/ConfigWriter.php [hunk 3]
```diff
@@ -88,7 +88,7 @@ public function saveConfigElement(Plugin $plugin, $name, $value, Shop $shop)
             throw new \Exception(sprintf('Config element "%s" not found.', $name));
         }
 
-        if ($element->getScope() == 0 && $shop->getId() !== 1) {
+        if ($element->getScope() === 0 && $shop->getId() !== 1) {
```
Comment (mitelg) [inline engine/Shopware/Components/Plugin/ConfigWriter.php]: don't we have a constant for the scopes?

File: engine/Shopware/Components/MultiEdit/Resource/Product/DqlHelper.php [hunk 4]
```diff
@@ -780,7 +780,7 @@ public function formatValue($prefix, $field, $value)
         $info = $this->getColumnsForProductListing();
         $info = $info[ucfirst($prefix) . ucfirst($field['field'])];
 
-        if ($info['nullable'] && $value == '') {
+        if ($info['nullable'] && $value === '') {
```
Comment (mitelg) [inline engine/Shopware/Components/MultiEdit/Resource/Product/DqlHelper.php:783]: did you checked this? if $value was false this  matched

File: engine/Shopware/Components/Document.php [hunk 5]
```diff
@@ -748,22 +756,22 @@ protected function saveDocument()
...
-            if ($typID != 4) {
+            if ($typID !== Document::TYPE_CANCEL) {
                 if (!empty($this->_document['numbers'])) {
-                    $numberrange = $this->_document['numbers'];
+                    $numberRange = $this->_document['numbers'];
                 } else {
                     // The typID is indexed with base 0, so we need increase the typID
-                    if (!in_array($typID, ['1', '2', '3'])) {
-                        $typID = $typID + 1;
+                    if (!in_array($typID, [1, 2, 3], true)) {
```
Comment (mitelg) [inline engine/Shopware/Components/Document.php]: why don't you use constants here?

File: engine/Shopware/Components/Document.php [hunk 6]
```diff
@@ -677,7 +685,7 @@ protected function saveDocument()
             WHERE `type` = ? AND userID = ? AND orderID = ? LIMIT 1
             ';
             $amount = ($this->_order->order->taxfree ? true : $this->_config['netto']) ? round($this->_order->amountNetto, 2) : round($this->_order->amount, 2);
-            if ($typID == 4) {
+            if ($typID === 4) {
```
Comment (mitelg) [inline engine/Shopware/Components/Document.php]: constant

File: engine/Shopware/Components/Api/Resource/PropertyGroup.php [hunk 7]
```diff
@@ -214,7 +220,7 @@ private function preparePropertyData(array $params, $propertyGroup = null)
             }
 
             // Sortmode equals the old article_count sorting?
-            if ($params['sortmode'] == 2) {
+            if ((int) $params['sortmode'] === 2) {
```
Comment (mitelg) [inline engine/Shopware/Components/Api/Resource/PropertyGroup.php]: are there constants for the sort mode?

File: engine/Shopware/Components/Plugin/ConfigWriter.php (hunk 3 above)
Comment (pascalheidmann) [reply engine/Shopware/Components/Plugin/ConfigWriter.php]: ~No, sadly not. But I might introduce it... ;)~
I found them in _Element_ model and will use them

File: engine/Shopware/Components/Api/Resource/PropertyGroup.php (hunk 7 above)
Comment (pascalheidmann) [reply engine/Shopware/Components/Api/Resource/PropertyGroup.php]: No, but actually I haven't found anything related to sortmode.
What are those actually values meant for?

File: engine/Shopware/Components/MultiEdit/Resource/Product/DqlHelper.php (hunk 4 above)
Comment (pascalheidmann) [reply engine/Shopware/Components/MultiEdit/Resource/Product/DqlHelper.php:783]: The function's doc block explicitly says
_Will set value = null for empty *strings* and replace comma with period for decimals_.
Therefore following its specs that should be ok I think.

Also the only usage of that function is in `Shopware\Components\MultiEdit\Resource\Product::save()` which itself gets its parameters via Request params which are always strings.

File: engine/Shopware/Components/Api/Resource/PropertyGroup.php (hunk 7 above)
Comment (mitelg) [reply engine/Shopware/Components/Api/Resource/PropertyGroup.php]: maybe @shyim has an idea

Comment (pascalheidmann) [reply engine/Shopware/Components/Api/Resource/PropertyGroup.php]: I found 3 consts in `Shopware\Bundle\StoreFrontBundle\Struct\Property\Set`, but those are values `0`, `1` and `3`.

**Pull request closed without merging**